pkg/datapath/sockets/sockets.go

```diff
@@ -0,0 +1,308 @@
+	sizeofSocketRequest = sizeofSocketID + 0x8
+	sizeofSocket        = sizeofSocketID + 0x18
+	// TODO: upstream
+	SOCK_DESTROY = 21
+)
+
+var (
+	log          = logging.DefaultLogger.WithField(logfields.LogSubsys, "datapath-sockets")
+	native       = nl.NativeEndian()
+	networkOrder = binary.BigEndian
+)
+
+const (
+	NETNS_ALL     = iota // Default: check for sockets in all network namespaces
+	NETNS_HOST           // Check for sockets only in the host network namespace
+	NETNS_NONHOST        // Check for sockets in non-host network namespaces
+)
+
+type SocketsDestroyer interface {
+	Destroy(filter SocketFilter) error
+}
+
+type SocketFilter struct {
+	DestIp   net.IP
+	DestPort uint16
+	Family   uint8
+	Protocol uint8
+	NetNs    uint8
+	// Optional callback function to determine whether a filtered socket needs to be destroyed
+	DestroyCB DestroySocketCB
+}
+
+type DestroySocketCB func(id netlink.SocketID) bool
+
+// Destroy destroys sockets matching the passed filter parameters using the
+// sock_diag netlink framework.
+//
+// Supported families in the filter: syscall.AF_INET, sycall.AF_INET6
+// Supported protocols in the filter: unix.IPPROTO_TCP, unix.IPPROTO_UDP
+func Destroy(filter SocketFilter) error {
+	family := filter.Family
+	protocol := filter.Protocol
+
+	if family != syscall.AF_INET && family != syscall.AF_INET6 {
+		return fmt.Errorf("unsupported family for socket destroy: %d", family)
+	}
+	var errs error
+	success, failed := 0, 0
+
+	// Query sockets matching the passed filter, and then destroy the filtered
+	// sockets.
+	switch protocol {
+	case unix.IPPROTO_TCP:
+		sockets, err := netlink.SocketDiagTCP(family)
+		if err != nil {
+			return fmt.Errorf("destroy failed: unable to get sockets with filter %v: %w", filter, err)
+		}
+		for _, sock := range sockets {
+			if sock.ID.Destination.Equal(filter.DestIp) && sock.ID.DestinationPort == filter.DestPort {
+				if filter.DestroyCB != nil && !filter.DestroyCB(sock.ID) {
+					continue
+				}
+				if err := destroySocket(sock.ID, family, unix.IPPROTO_TCP); err != nil {
+					errs = errors.Join(errs, fmt.Errorf("TCP socket with filter [%v]: %w", filter, err))
+					failed++
+					continue
+				}
+				log.Debugf("Destroyed socket: %v", sock.ID)
+				success++
+			}
+		}
+	case unix.IPPROTO_UDP:
+		sockets, err := socketDiagUDP(family)
+		if err != nil {
+			return fmt.Errorf("destroy failed: unable to get sockets with filter %v: %w", filter, err)
+		}
+		log.Info("BEFORE")
+		for _, sock := range sockets {
+			if sock.ID.Destination.Equal(filter.DestIp) && sock.ID.DestinationPort == filter.DestPort {
+				if filter.DestroyCB != nil && !filter.DestroyCB(sock.ID) {
+					continue
+				}
+				if err := destroySocket(sock.ID, family, unix.IPPROTO_UDP); err != nil {
+					errs = errors.Join(errs, fmt.Errorf("UDP socket with filter [%v]: %w", filter, err))
+					failed++
+					continue
+				}
+				log.Debugf("Destroyed socket: %v", sock.ID)
+				success++
+			}
+		}
```

Comment:
we may add a TODO for SCTP? it is GA in kubernetes and cilium has some functionality already

Reply:
SCTP isn't a supported protocol with cilium's kube-proxy replacement, it's documented.